### PR TITLE
Use Java 8 variant of Guava (take 2)

### PR DESCRIPTION
**Pull request by lptr**
Let's try doing #8991 again.

Depends on https://github.com/gradle/gradle/pull/9702

### CODE REVIEW

**Comment by lptr** [comment]
The performance regression is there, but it only affects native monolithic projects: https://builds.gradle.org/viewLog.html?buildId=23124382&buildTypeId=Gradle_Check_PerformanceTestCoordinator&tab=report_project944_Performance&branch_Gradle_Check_Stage_ReadyforMerge=release

This makes sense as `CachingTaskInputFileCollection` where the regression occurs is only used in native compilation currently.

**Comment by lptr** [comment]
@paplorinc is going to try to mitigate the regression today, but if he can't find anything simple to do that he can implement by end of day, I think we should accept the regression and move on with our lives.

**Comment by lptr** [comment]
Let's try omitting `CachingTaskInputFileCollection` altogether and see if it solves the regression. We can address caching for file collections in a more wholistic way in a separate PR.

**Comment by lptr** [comment]
Removing the caching made things a lot worse. :(

```text
Speed Results for test project 'nativeMonolithic' with tasks build: we're slower than 5.6-commit-6c332fa with 99% confidence.
Difference: 4.128 s slower (4128.5 ms), 40.92%
  Current Gradle median: 14.218 s min: 14.088 s, max: 16.22 s, se: 620.14 ms}> [16.22 s, 14.306 s, 14.259 s, 14.166 s, 14.23 s, 14.814 s, 14.207 s, 14.123 s, 14.138 s, 14.088 s]
  Gradle 5.6-commit-6c332fa median: 10.09 s min: 9.998 s, max: 11.533 s, se: 436.21 ms}> [10.294 s, 11.533 s, 10.17 s, 10.086 s, 10.143 s, 10.056 s, 10.094 s, 9.998 s, 10.034 s, 10.055 s
```

See https://builds.gradle.org/viewQueued.html?itemId=23667354

**Comment by wolfs** [comment]
Shall we also try using an `immutableList` instead of `ImmutableSet`?

**Comment by lptr** [comment]
Right, that's the other thing we wanted to try. Coming in a bit...

**Comment by lptr** [comment]
Guava 28.0 is out with a potential fix for the regression: https://github.com/google/guava/commit/57e4b6d560874946d7142103386f7f17384e1708

I'm checking it out now.

**Comment by lptr** [comment]
Looks like our regression comes from another source:

```text
Speed Results for test project 'nativeMonolithicOverlapping' with tasks build: we're slower than 5.6-commit-c588a34 with 99% confidence.
Difference: 554.5 ms slower (554.5 ms), 5.67%
  Current Gradle median: 10.326 s min: 10.145 s, max: 11.251 s, se: 360.02 ms}
  > [10.469 s, 10.333 s, 10.339 s, 11.094 s, 10.319 s, 10.26 s, 10.234 s, 10.289 s, 11.251 s, 10.145 s]
  Gradle 5.6-commit-c588a34 median: 9.772 s min: 9.69 s, max: 11.209 s, se: 438.1 ms}
  > [9.972 s, 9.988 s, 11.209 s, 9.817 s, 9.79 s, 9.753 s, 9.701 s, 9.69 s, 9.69 s, 9.713 s]
```

https://builds.gradle.org/viewLog.html?buildId=23809186&buildTypeId=Gradle_Check_IndividualPerformanceScenarioWorkersLinux

**Comment by paplorinc** [comment]
I looked into the flames  and it *does* seem to me that the culprit is the new Set handling ([memory-wise](https://builds.gradle.org/repository/download/Gradle_Util_Performance_AdHocPerformanceScenarioLinux/23820888:id/results/performance/build/test-results-performanceAdhocTest.zip!/flames/5.6-commit-c588a34/allocation/simplified/diffs/flame-backward-diff-vs-current.svg) and [cpu-wise](https://builds.gradle.org/repository/download/Gradle_Util_Performance_AdHocPerformanceScenarioLinux/23820888:id/results/performance/build/test-results-performanceAdhocTest.zip!/flames/5.6-commit-c588a34/cpu/simplified/diffs/flame-backward-diff-vs-current.svg)):
Screen Shot 2019-06-12 at 19 07 27

Also, according to the [documentation](https://github.com/gradle/gradle/blob/1ee5ff831577fbccc4d8ef047eaff239fd72de26/subprojects/core/src/main/java/org/gradle/initialization/ParallelismBuildOptions.java#L76) `max-workers` cannot be 0, so I changed it to 1 (couldn't run it with a value of 0 from the command line either)

Interestingly there seems to be [extra IO](https://builds.gradle.org/repository/download/Gradle_Util_Performance_AdHocPerformanceScenarioLinux/23820888:id/results/performance/build/test-results-performanceAdhocTest.zip!/flames/5.6-commit-c588a34/io/raw/diffs/flame-backward-diff-vs-current.svg) as well, not sure what it means.

File: subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/library-versions.properties [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
-guava=27.1-jre
+guava=28.0-jre
 junit-jupiter=5.4.2
 junit=4.12
 kotlin=1.3.31
 scala-library=2.12.8
 scala-xml=1.2.0
 scala=2.12
-scalatest=3.0.7
+scalatest=3.0.8
```
Comment (paplorinc) [inline subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/library-versions.properties]: `updateInitPluginTemplateVersionFile` changed this as well

Comment (lptr) [reply subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/library-versions.properties]: These should be updated automatically during release.

File: subprojects/docs/src/samples/userguide/dependencyManagement/artifactTransforms/minify/groovy/build.gradle [hunk 2]
```diff
@@ -142,7 +142,7 @@ repositories {
 // tag::artifact-transform-attribute-setup[]
 
 dependencies {                                 // <4>
-    implementation('com.google.guava:guava:27.1-jre')
+    implementation('com.google.guava:guava:28.0-jre')
```
Comment (lptr) [inline subprojects/docs/src/samples/userguide/dependencyManagement/artifactTransforms/minify/groovy/build.gradle]: This is the same case as with Groovy: we shouldn't upgrade samples, instead we should figure out a way to do this automatically.

File: subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/library-versions.properties (hunk 1 above)
Comment (paplorinc) [reply subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/library-versions.properties]: k, reverted

File: subprojects/docs/src/samples/userguide/dependencyManagement/artifactTransforms/minify/groovy/build.gradle (hunk 2 above)
Comment (paplorinc) [reply subprojects/docs/src/samples/userguide/dependencyManagement/artifactTransforms/minify/groovy/build.gradle]: reverted

**Comment by lptr** [comment]
OK, so we are back to the drawing board. Let's retry moving to Java 8 after @adammurdoch had a chance to introduce caching/memoization for all file collections in the coming weeks.

**Comment by lptr** [comment]
@paplorinc can you please remove the last commit?

**Comment by paplorinc** [comment]
> so we are back to the drawing board

Can I ask what makes you say that? I don't understand the build failure, the test passes locally.

**Comment by paplorinc** [comment]
@lptr, the changes did seem to make the situation slightly better, i.e. the failing scenarios are within boundaries now:
Screen Shot 2019-06-14 at 15 52 55

It could obviously be just a fluke, but I won't spend time on it, as agreed.
Do you still want me to remove the change?

**Comment by paplorinc** [comment]
@lptr, Ready for Merge passed (and Ready for Release fails on master as well).
Please take over from here.

**Comment by lptr** [comment]
Looks like the recent changes in https://github.com/gradle/gradle/pull/9766 didn't fix the regression yet.

**Comment by paplorinc** [comment]
What about the reverted commit, the build has passed with those changes?

**Comment by lptr** [comment]
Closing in favor of https://github.com/gradle/gradle/pull/12472

**Pull request closed without merging**